File: app/src/main/java/org/wikipedia/feed/news/NewsCardView.java
package org.wikipedia.feed.news;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.PagerSnapHelper;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SnapHelper;

import org.wikipedia.R;
import org.wikipedia.feed.view.CardHeaderView;
import org.wikipedia.feed.view.DefaultFeedCardView;
import org.wikipedia.feed.view.FeedAdapter;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.ResourceUtil;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static org.wikipedia.util.L10nUtil.isLangRTL;

public class NewsCardView extends DefaultFeedCardView<NewsCard> {
    @BindView(R.id.news_recycler_view) RecyclerView newsRecyclerView;
    @BindView(R.id.header_view) CardHeaderView headerView;
    @BindView(R.id.rtl_container) View rtlContainer;
    private boolean isSnapHelperAttached;

    public interface Callback {
        void onNewsItemSelected(@NonNull NewsCard card, NewsItemView view);
    }

    public NewsCardView(@NonNull Context context) {
        super(context);
        inflate(getContext(), R.layout.view_card_news, this);
        ButterKnife.bind(this);
    }

    private void setUpIndicatorDots() {
        final int indicatorRadius = 4;
        final int indicatorPadding = 8;
        final int indicatorHeight = 20;
        newsRecyclerView.addItemDecoration(new RecyclerViewIndicatorDotDecor(DimenUtil.roundedDpToPx(indicatorRadius),
                DimenUtil.roundedDpToPx(indicatorPadding), DimenUtil.roundedDpToPx(indicatorHeight),
                ResourceUtil.getThemedColor(getContext(), R.attr.chart_shade5),
                ResourceUtil.getThemedColor(getContext(), R.attr.colorAccent), isLangRTL(getCard().wikiSite().languageCode())));
    }

    @Override
    public void setCallback(@Nullable FeedAdapter.Callback callback) {
        super.setCallback(callback);
        headerView.setCallback(callback);
    }

    @Override public void setCard(@NonNull NewsCard card) {
        if (card.equals(getCard())) {
            return;
        }
        super.setCard(card);
        header(card);
        setLayoutDirectionByWikiSite(card.wikiSite(), rtlContainer);
        setUpRecycler(card);
    }

    @SuppressWarnings("checkstyle:magicnumber")
    private void setUpRecycler(NewsCard card) {
        newsRecyclerView.setHasFixedSize(true);
        newsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        newsRecyclerView.setNestedScrollingEnabled(false);
        newsRecyclerView.setClipToPadding(false);
        newsRecyclerView.setAdapter(new NewsAdapter(card));
        setUpIndicatorDots();
        setUpSnapHelper();
    }

    private void setUpSnapHelper() {
        if (!isSnapHelperAttached) {
            final SnapHelper snapHelper = new PagerSnapHelper();
            snapHelper.attachToRecyclerView(newsRecyclerView);
            isSnapHelperAttached = true;
        }
    }

    private void header(@NonNull NewsCard card) {
        headerView.setTitle(card.title())
                .setLangCode(card.wikiSite().languageCode())
                .setCard(card);
    }

    private static class NewsItemHolder extends RecyclerView.ViewHolder {
        NewsItemView itemView;

        NewsItemHolder(NewsItemView itemView) {
            super(itemView);
            this.itemView = itemView;
        }

        void bindItem(NewsItem newsItem) {
            itemView.setContents(newsItem);
        }

        NewsItemView getView() {
            return itemView;
        }
    }

    private class NewsAdapter extends RecyclerView.Adapter<NewsItemHolder> {
        private List<NewsItem> newsItems = new ArrayList<>();
        private NewsCard card;

        NewsAdapter(@NonNull NewsCard card) {
            this.card = card;
            this.newsItems.addAll(card.news());
        }

        @Override
        public int getItemCount() {
            return newsItems.size();
        }

        @NonNull
        @Override
        public NewsItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new NewsItemHolder(new NewsItemView(getContext()));
        }

        @Override
        public void onBindViewHolder(@NonNull NewsItemHolder holder, int position) {
            holder.bindItem(newsItems.get(position));
            holder.getView().setOnClickListener(view -> {
                if (getCallback() != null) {
                    getCallback().onNewsItemSelected(card, holder.getView());
                }
            });
        }

        @Override
        public void onViewAttachedToWindow(@NonNull NewsItemHolder holder) {
            super.onViewAttachedToWindow(holder);
            holder.getView().setCallback(getCallback());
        }

        @Override
        public void onViewDetachedFromWindow(@NonNull NewsItemHolder holder) {
            holder.getView().setCallback(null);
            super.onViewDetachedFromWindow(holder);
        }
    }
}

File: app/src/main/java/org/wikipedia/feed/news/NewsCardView.kt
package org.wikipedia.feed.news

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import org.wikipedia.R
import org.wikipedia.databinding.ViewCardNewsBinding
import org.wikipedia.feed.view.DefaultFeedCardView
import org.wikipedia.feed.view.FeedAdapter
import org.wikipedia.settings.Prefs
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.L10nUtil
import org.wikipedia.util.ResourceUtil
import org.wikipedia.views.ImageZoomHelper

class NewsCardView(context: Context) : DefaultFeedCardView<NewsCard>(context) {

    interface Callback {
        fun onNewsItemSelected(card: NewsCard, view: NewsItemView)
    }

    private val binding = ViewCardNewsBinding.inflate(LayoutInflater.from(context), this, true)
    private var prevImageDownloadSettings = Prefs.isImageDownloadEnabled()
    private var isSnapHelperAttached = false

    private fun setUpIndicatorDots(card: NewsCard) {
        val indicatorRadius = 4
        val indicatorPadding = 8
        val indicatorHeight = 20
        binding.newsRecyclerView.addItemDecoration(
            RecyclerViewIndicatorDotDecor(
                DimenUtil.roundedDpToPx(indicatorRadius.toFloat()).toFloat(),
                DimenUtil.roundedDpToPx(indicatorPadding.toFloat()),
                DimenUtil.roundedDpToPx(indicatorHeight.toFloat()),
                ResourceUtil.getThemedColor(context, R.attr.chart_shade5),
                ResourceUtil.getThemedColor(context, R.attr.colorAccent),
                L10nUtil.isLangRTL(card.wikiSite().languageCode)
            )
        )
    }

    override var callback: FeedAdapter.Callback? = null
        set(value) {
            field = value
            binding.headerView.setCallback(value)
        }

    override var card: NewsCard? = null
        set(value) {
            if (field != value || prevImageDownloadSettings != Prefs.isImageDownloadEnabled()) {
                field = value
                prevImageDownloadSettings = Prefs.isImageDownloadEnabled()
                value?.let {
                    header(it)
                    setLayoutDirectionByWikiSite(it.wikiSite(), binding.rtlContainer)
                    setUpRecycler(it)
                }
            }
        }

    private fun setUpRecycler(card: NewsCard) {
        binding.newsRecyclerView.setHasFixedSize(true)
        binding.newsRecyclerView.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        binding.newsRecyclerView.isNestedScrollingEnabled = false
        binding.newsRecyclerView.clipToPadding = false
        binding.newsRecyclerView.adapter = NewsAdapter(card)
        setUpIndicatorDots(card)
        setUpSnapHelper()
    }

    private fun setUpSnapHelper() {
        if (!isSnapHelperAttached) {
            val snapHelper = PagerSnapHelper()
            snapHelper.attachToRecyclerView(binding.newsRecyclerView)
            isSnapHelperAttached = true
        }
    }

    private fun header(card: NewsCard) {
        binding.headerView.setTitle(card.title())
            .setLangCode(card.wikiSite().languageCode)
            .setCard(card)
    }

    private class NewsItemHolder constructor(private val newsItemView: NewsItemView) : RecyclerView.ViewHolder(newsItemView) {
        fun bindItem(newsItem: NewsItem) {
            newsItemView.setContents(newsItem)
        }

        val view get() = newsItemView
    }

    private inner class NewsAdapter constructor(private val card: NewsCard) : RecyclerView.Adapter<NewsItemHolder>() {

        override fun getItemCount(): Int {
            return card.news().size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsItemHolder {
            return NewsItemHolder(NewsItemView(context))
        }

        override fun onBindViewHolder(holder: NewsItemHolder, position: Int) {
            holder.bindItem(card.news()[position])
            holder.view.setOnClickListener {
                callback?.onNewsItemSelected(card, holder.view)
            }
            holder.view.setOnLongClickListener {
                if (ImageZoomHelper.isZooming) {
                    ImageZoomHelper.dispatchCancelEvent(holder.view)
                }
                true
            }
        }

        override fun onViewAttachedToWindow(holder: NewsItemHolder) {
            super.onViewAttachedToWindow(holder)
            holder.view.callback = callback
        }

        override fun onViewDetachedFromWindow(holder: NewsItemHolder) {
            holder.view.callback = null
            super.onViewDetachedFromWindow(holder)
        }
    }
}
